org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/semanticHighlight/SemanticHighlightingRegistry.xtend

```diff
@@ -0,0 +1,220 @@
+		return if(index === null) -1 else index;
+	}
+
+	/**
+	 * Returns with a view of all scopes known by this manager.
+	 */
+	def List<List<String>> getAllScopes() {
+		checkInitialized;
+		val builder = ImmutableList.builder;
+		scopes.keySet.forEach [
+			builder.add(Preconditions.checkNotNull(scopes.get(it), '''No scopes are available for index: �it�'''))
+		];
+		return builder.build;
+	}
+
+	def void update(ILanguageServerAccess.Context context) {
+		checkInitialized;
+		if (!(context.resource instanceof XtextResource)) {
+			return;
+		}
+		if (!context.documentOpen) {
+			return;
+		}
+
+		val resource = context.resource as XtextResource;
+		val calculator = resource.resourceServiceProvider?.get(ISemanticHighlightingCalculator);
+		val mapper = resource.resourceServiceProvider?.get(ISemanticHighlightingStyleToTokenMapper);
+		if (calculator === null || mapper === null) {
+			return;
+		}
+
+		val document = context.document;
+		val acceptor = new MergingHighlightedPositionAcceptor(calculator);
+		calculator.provideHighlightingFor(resource, acceptor, CancelIndicator.NullImpl);
+		val ranges = acceptor.positions.map [ position |
+			position.ids.map [ id |
+				val start = document.getPosition(position.offset);
+				val end = document.getPosition(position.offset + position.length);
+				val scope = getIndex(mapper.toScopes(id));
+				return new HighlightedRange(start, end, scope);
+			]
+		].flatten;
+
+		val lines = ranges.toLines(document);
+		val textDocument = context.toTextDocument;
+		notifyClient(new SemanticHighlightingParams(textDocument, lines));
+	}
+
+	protected def List<SemanticHighlightingInformation> toLines(Iterable<? extends HighlightedRange> ranges, Document document) {
+		val builder = ImmutableMultimap.builder;
+		ranges.filter[start != end].forEach [
+			if (start.line === end.line) {
+				val length = end.character - start.character;
+				builder.put(start.line, new SemanticHighlightingTokens.Token(start.character, length, scope));
+			} else {
+				val startLine = start.line;
+				val endLine = end.line;
+				val startLineContent = document.getLineContent(startLine);
+				val startLength = startLineContent.length - start.character;
+				builder.put(startLine, new SemanticHighlightingTokens.Token(start.character, startLength, scope));
+				builder.put(endLine, new SemanticHighlightingTokens.Token(0, end.character, scope));
+				for (var line = startLine + 1; line < end.line; line++) {
+					val lineContent = document.getLineContent(line);
+					builder.put(line, new SemanticHighlightingTokens.Token(0, lineContent.length, scope));
+				}
+			}
+		];
+		return builder.build.asMap.entrySet.map [
+			val line = key;
+			val tokens = value;
+			new SemanticHighlightingInformation(line, tokens.encode);
+		].toList;
+	}
+
+	protected def VersionedTextDocumentIdentifier toTextDocument(ILanguageServerAccess.Context context) {
+		return new VersionedTextDocumentIdentifier => [
+			uri = context.resource.URI.toUriString;
```

Comment:
do you know why the context does not store the original uri. if yes we could move this as member to context

Reply:
No, I have no idea.